src/lavinmqperf/mqtt/throughput.cr

```diff
@@ -0,0 +1,345 @@
+
+      def initialize
+        super
+        @parser.on("-x publishers", "--publishers=number", "Number of publishers (default 1)") do |v|
+          @publishers = v.to_i
+        end
+        @parser.on("-y consumers", "--consumers=number", "Number of consumers (default 1)") do |v|
+          @consumers = v.to_i
+        end
+        @parser.on("-s msgsize", "--size=bytes", "Size of each message (default 16 bytes)") do |v|
+          @size = v.to_i
+        end
+        @parser.on("-V", "--verify", "Verify the message body") do
+          @verify = true
+        end
+        @parser.on("-q qos", "--qos=level", "QoS level (0 or 1)") do |v|
+          @qos = v.to_i
+        end
+        @parser.on("-t topic", "--topic=name", "Topic name (default perf-test)") do |v|
+          @topic = v
+        end
+        @parser.on("-r pub-rate", "--rate=number", "Max publish rate (default 0)") do |v|
+          @rate = v.to_i
+        end
+        @parser.on("-R consumer-rate", "--consumer-rate=number", "Max consume rate (default 0)") do |v|
+          @consume_rate = v.to_i
+        end
+        @parser.on("-j", "--json", "Output result as JSON") do
+          @json_output = true
+        end
+        @parser.on("-z seconds", "--time=seconds", "Only run for X seconds") do |v|
+          @timeout = Time::Span.new(seconds: v.to_i)
+        end
+        @parser.on("-q", "--quiet", "Quiet, only print the summary") do
+          @quiet = true
+        end
+        @parser.on("-C messages", "--pmessages=messages", "Publish max X number of messages") do |v|
+          @pmessages = v.to_i
+        end
+        @parser.on("-D messages", "--cmessages=messages", "Consume max X number of messages") do |v|
+          @cmessages = v.to_i
+        end
+        @parser.on("--random-bodies", "Each message body is random") do
+          @random_bodies = true
+        end
+        @parser.on("--retain", "Set retain flag on published messages") do
+          @retain = true
+        end
+        @parser.on("--clean-session", "Use clean session") do
+          @clean_session = true
+        end
+        @parser.on("-u uri", "--uri=uri", "MQTT broker URI (default mqtt://localhost:1883)") do |v|
+          @uri = URI.parse(v)
+        end
+      end
+
+      @pubs = 0_u64
+      @consumes = 0_u64
+      @stopped = false
+      @counter_lock = Mutex.new
+
+      private def create_client(id : Int32, role : String) : {TCPSocket, LavinMQ::MQTT::IO}
+        host = @uri.host || "localhost"
+        port = @uri.port || 1883
+        user = @uri.user || "guest"
+        password = @uri.password || "guest"
+
+        socket = TCPSocket.new(host, port)
+        socket.keepalive = true
+        socket.tcp_nodelay = false
+        socket.sync = true
+        io = LavinMQ::MQTT::IO.new(socket)
+
+        client_id = "#{role}-#{id}"
+        connect_packet = LavinMQ::MQTT::Connect.new(
+          client_id: client_id,
+          clean_session: @clean_session,
+          keepalive: 30_u16,
```

Comment:
Maybe an improvement for later but configure keepalive using query params like `-u mqtt://hostname?keepalive=10`